### ISSUE TITLE
typescript port avaible

### ISSUE THREAD

**Issue by tx4x** [szwacz/fs-jetpack#40]
hi,
as mentioned, I have to modify fs-jetpack in a way it fits to my needs. I ported all to Typescript, see https://github.com/xblox/fs. I am not sure how long this repo will be there, so if there is any interest, you may make a backup. 

this was just for your information, so close this here just
have fun,
g

**Comment by xblox** [comment]
[this](https://github.com/xblox/fs/blob/master/typings.d.ts) are the typings for your library (needs correction on your side) you can upload to 'definitely typed' but also just append your package.json for 'typings':'./typings.dts' (so people don't need to use the dtyped registry). that makes TS users happy, always.
thanks again & enjoy yourself
guenter

**Comment by szwacz** [comment]
Thanks @xblox. I'll have those typings in mind. But will postpone implementing them until more people will ask for them. After all the sanest thing to do for typescripters would be to use your fork.

**Comment by aurbano** [comment]
This would be quite helpful actually :)

**Comment by ashleydavis** [comment]
I'd love to have TypeScript typings for fs-jetpack.

**Comment by szwacz** [comment]
Ok, I'll see what I can do :)

**Comment by astoilkov** [comment]
We heavily depend on this great library. However, we are starting to move all our codebase to TypeScript so type definitions would be awesome.

Thanks in advance.

**Comment by szwacz** [comment]
I did some introductory thinking and it's going to happen. But I don't want to prepare a typings file, I want to rewrite the whole codebase to TypeScript.

Please ping me here if it's not done within 3 weeks, calling me a lazy slob ;)

**Comment by tx4x** [comment]
@szwacz, i've done the most of the typescript port as said : https://github.com/tx4x/fs. i have added a couple more things: copy, move, delete progress, user callbacks for conflicts. it was a while ago so your new things are not there :-)
g

**Comment by szwacz** [comment]
All right ladies and gents! We're almost done.
The code isn't published yet, because I want to discuss with you one design decission (I'm not active user of TypeScript with Node).
Here is the explanation of what I did: https://github.com/szwacz/fs-jetpack/tree/ts-typings#usage-with-typescript
And here are two typing files:
1: https://github.com/szwacz/fs-jetpack/blob/ts-typings/types.ts
2: https://github.com/szwacz/fs-jetpack/blob/ts-typings/index.d.ts
Are there any node.js & TypeScript conventions I'm not aware of?

**Comment by astoilkov** [comment]
I think you should take a look at how [minimist](https://github.com/substack/minimist) does this. It is written in JavaScript and there aren't any TypeScript files inside of the module. However, when you install `@types/minimist` which is the universal way of adding TypeScript definitions for JavaScript projects you get the `minimist()` function and `minimist` namespace at the same time. Look at the file [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/minimist/index.d.ts).

**Comment by astoilkov** [comment]
@szwacz Ok. I get it now. Then your solution seems a good idea. I have no more suggestions. One quick question - Why did you decide to write the TypeScript definitions not rewrite the library in TypeScript?

**Comment by szwacz** [comment]
Typings are shipped in `v2.1.0` 🎉 

@astoilkov I've started with full rewrite into TypeScript, but was pacing very slowly. This libaray has a lot of `fn(...args)` stuff and in some cases I've struggled to write it as a correct TypeScript code. So after some time with very little progress just decided that typings file is at least something I can ship in reasonable time :)

The tests are rewritten into typescript as well, so the typings file is tested, and should be of high quality.
